designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,544 @@
+
+```js
+const fs = require("fs");
+
+exports.config = {
+    files: "*.js",
+    ignore: fs.readFileSync(".eslintignore").replace(/\n/g, ";")
+};
+```
+
+### The `@eslint/config` Package
+
+Because some of the operations ESLint currently do are quite complicated, this design includes a utility package called `@eslint/config` that users can use for common tasks. This package contains the following utilities:
+
+1. `Config` - a class that both wraps a configuration object to provide helper methods for manipulating configs and a container for static methods to work with configs.
+1. `RuleLoader` - a class that aids in loading rules from a directory.
+
+The `@eslint/config` package is intended to add back functionality that would be removed from ESLint with this design.
+
+**Note:** It is not required to use this package with `.eslint.js`; this is an optional tool that users may choose to use for convenience.
+
+#### `Config` Class
+
+The `Config` class has this form:
+
+```js
+class Config {
+
+    // pass in any object literal to wrap in Config
+    constructor(object) {}
+
+    // convenience method for easily setting rule severity
+    setRuleSeverity(ruleId, severity) {}
+
+    // determine if an object is a Config
+    static isConfig(object) {}
+
+    // create a new config based on another config
+    static create(baseConfig, overrideConfig) {}
+
+    // merge info from other configs into a base config
+    static assign(receiverConfig, ...supplierConfigs) {}
+}
+```
+
+A new instance of `Config` is created like this:
+
+```js
+const { Config } = require("@eslint/config");
+
+const config = new Config({
+    files: "*.js",
+    ruledefs: {
+        react: require("eslint-plugin-react").rules,
+    },
+    rules: {
+        "react/jsx-uses-react": "error",
+        semi: ["error", "always"]
+    }
+});
+```
+
+After that point, you can easily change the severity of a rule like this:
+
+```js
+config.setRuleSeverity("semi", "warn");
+```
+
+In this way, users don't need to modify an array to change a rule's severity, which is a major downside of this design when using a vanilla object structure as a config.
+
+The static methods on `Config` make it easy to create and modify configs. For example, suppose you want to create a config based on `eslint-config-standard` but would prefer `semi` be a warning instead of an error, you can do so using `Config.create()`:
+
+```js
+const standardConfig = require("eslint-config-standard");
+const { Config } = require("@eslint/config");
+
+const myConfig = Config.create(standardConfig, {
+    rules: {
+        semi: "warn"
+    }
+});
+
+exports.config = myConfig;
+```
+
+Here, the `Config.create()` method creates a new `Config` object based on `standardConfig` and then merges in the overrides from the second argument. The overrides config just changed the severity of `semi` to be `"warn"` but does not affect the other `semi` options. (This implements the current config merging functionality in ESLint.)
```

Comment:
Yup. It's solved the exact same way. The last one wins.

Reply:
I'm not sure I understand. What would be the result of merging two configs like this?

```js
// config 1
module.exports = {
  ruledefs: {
    foo: {
      "my-rule1": require("rule-1"),
      "my-rule2": require("rule-2")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option1"]
    "foo/my-rule2": "error"
  }
};
```

```js
// config 2
module.exports = {
  ruledefs: {
    foo: {
      "my-rule1": require("some-other-rule1")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option2"]
  }
};
```

```js
Config.create(config1, config2) // => ?
```

Both configs are individually valid, but it's not clear how merging the configs would create another valid config. (Based on the "last one wins" rule, it seems like this would result in a config that (a) references a rule called `foo/my-rule2` even though no definition for such a rule is provided, and (b) provides the `"option1"` option to the rule at `require("some-other-rule1")`, which might not be a valid option for that rule.)

I'm not sure what you mean by "it's solved the exact same way" because I don't think we've solved this problem before, aside from the solution in #5.

Reply:
"It's solved the same way" -> Even with `.eslintc`, it's the last config whose settings win. So in an `extends` array, the last config's settings override everything that came before it when there is a conflict, and then the config with `extends` takes priority over everything we calculate from `extends`.

`Config.create(a, b)` effectively creates `merge({ extends: a }, b)` (where `merge()` is our current merging strategy). So in your example, the resulting config would be:

```js
exports.config = {
    extends: config1,
  ruledefs: {
    foo: {
      "my-rule1": require("some-other-rule1")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option2"]
  }
};
```

Or, if we hydrate the config to remove `extends`, the resulting config would look like this:

```js
// config 1
exports.config = {
  ruledefs: {
    foo: {
      "my-rule1": require("some-other-rule1")
      "my-rule2": require("rule-2")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option2"]
    "foo/my-rule2": "error"
  }
};
```

Reply:
I see. What about this example?

<details>
<summary>Example</summary>

```js
// config 1
module.exports = {
  ruledefs: {
    foo: {
      "my-rule1": require("rule-1"),
      "my-rule2": require("rule-2")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option1"]
    "foo/my-rule2": "error"
  }
};
```

```js
// config 2
module.exports = {
  ruledefs: {
    foo: {
      "my-rule1": require("some-other-rule1"),
      "my-rule2": require("some-other-rule2")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option2"]
  }
};
```

```js
module.exports.config = { extends: [config1, config2] }; // => What is the flattened equivalent?
```

</details>

The only difference between this example and the previous example is that config 2 now defines a `foo/my-rule2` rule, but does not configure it.

(Note: The existing merging strategy did not need to handle cases like this, because we required plugin names to be globally unique.)

I also notice that in the final merged config, the ruledef for `foo` contains two rules (`my-rule1` and `my-rule2`) from two different sources (one from each config). I'm concerned that this would create a confusing debugging experience for users; for example, if they need to report a bug to `eslint-plugin-foo`, it might not be obvious which version a particular rule is coming from. (Additionally, since exporting ruledefs under the name `foo` would no longer require someone to control the `eslint-plugin-foo` package on npm, it seems like having different plugins with the same name would become more common in practice.)

Reply:
The existing merge strategy is that any top-level key that contains an object value has its properties merged, with the last one winning whenever those properties conflict. That's the same approach we'd use in this design for `ruledefs` (same merging strategy as `globals`, for example).

For this example, the hydrated result would be:

```js
// config 1
module.exports = {
  ruledefs: {
    foo: {
      "my-rule1": require("some-other-rule1"),
      "my-rule2": require("some-other-rule2")
    }
  },
  rules: {
    "foo/my-rule1": ["error", "option2"]
    "foo/my-rule2": "error"
  }
};
```

> Additionally, since exporting ruledefs under the name foo would no longer require someone to control the eslint-plugin-foo package on npm, it seems like having different plugins with the same name would become more common in practice.

As I've mentioned elsewhere, this is one of the downsides of this design: we lose control over package names and structures so we can no longer enforce anything. A lot of the issues you describe will have to be handled through documentation now. That's also why we need something like `ruledefs` so the end user can define the namespace for rules on their own.

I think we can mitigate some of this by encouraging people to add URLs to rule meta information and creating a new core formatter that outputs those URLs in some meaningful way.

Reply:
> For this example, the hydrated result would be:
> 
> ```js
> // config 1
> module.exports = {
>   ruledefs: {
>     foo: {
>       "my-rule1": require("some-other-rule1"),
>       "my-rule2": require("some-other-rule2")
>     }
>   },
>   rules: {
>     "foo/my-rule1": ["error", "option2"]
>     "foo/my-rule2": "error"
>   }
> };
> ```

Thanks for clarifying. This seems like a major problem to me. The config `"foo/my-rule2": "error"` was a configuration for the rule at `require("rule-2")`, but after configs get merged, that rule configuration ends up getting applied to a different rule at `require("some-other-rule2")`. If the two rules have incompatible schemas, this could result in a confusing config error when extending two different shareable configs.

In other words, I don't think this proposal currently offers a solution to the issues discussed in https://github.com/eslint/eslint/issues/3458 where two shareable configs depend on different versions of a plugin.

---

I see a few possible workarounds. In increasing order of complexity:

1. Raise a fatal error when merging two configs if they have overlapping `ruledefs` keys. Add a utility function to `@eslint/config` that renames a particular ruledef key in a config to something else, so that the end user can avoid these conflicts manually when merging configs. For example, an end-user's config might look like this:

    <details>
    <summary>Example config with solution 1</summary>

    ```js
    const { renameRuleDef } = require("@eslint/config");
    module.exports = {
        extends: [
            require("config1"),
            renameRuleDef(require("config2"), { from: "foo", to: "other-foo" })
        ],
        rules: {

            // configures the rule at require("rule-2")
            "foo/my-rule2": "error",

            // configures the rule at require("some-other-rule2")
            // (since that rule now appears in an "other-foo" ruledef)
            "other-foo/my-rule2": "error"
        }
    }
    ```
    </details>

    As a disadvantage, creating a config like this would probably be difficult for users who don't understand the details of how ruledefs work.

1. Introduce a rule-namespacing solution along the lines of https://github.com/eslint/rfcs/pull/5. We could probably avoid much of the complexity of #5 since this RFC removes so many features, but if we took this solution we might want to require users to name their `extends` clauses, so that users could refer to edges in the config tree. For example:

    <details>
    <summary>Example config with solution 2</summary>

    ```js
    module.exports = {
        // Note: `extends` is now an object, not an array
        extends: {
            config1: require("config1"),
            config2: require("config2")
        },
        rules: {

            // configures rules in the `foo` ruledef defined in `config1`
            "config1::foo/my-rule2": "error",

            // configures rules in the `foo` ruledef defined in `config2
            "config2::foo/my-rule2": "error",

            // If the "bar" ruledef is unambiguous, a `::` prefix is not required.
            "bar/my-rule3": "error"
        }
    }
    ```
    </details>

    As a disadvantage, this solution would require users to give explicit names to their extended configs (since `extends` would be an object) even in cases where no ambiguity was present. Note that this solution could not be added on after-the-fact without another big breaking change, since it would change the structure of `extends` clauses.

1. Combine the `ruledefs` and `rules` config options, so that users would load rules directly by symbols or object identity (similar to how users load parsers and processors in this proposal). This would avoid all naming conflicts because we wouldn't need to give rules user-declared names. For example:

    <details>
    <summary>Example config with solution 3</summary>

    ```js
    const fooPlugin = require("eslint-plugin-foo");
    module.exports = {
        rules: {
            [fooPlugin.rules["my-rule1"].KEY]: "error"
        }
    };
    ```
    </details>

    Solution 3 would probably introduce more complexity than it's worth, so I wouldn't recommend it at this point, but I'm mentioning it for the sake of completeness.

Reply:
> The config "foo/my-rule2": "error" was a configuration for the rule at require("rule-2"), but after configs get merged, that rule configuration ends up getting applied to a different rule at require("some-other-rule2").

That's a good point. Would it make sense to throw an error if `foo/my-rule-2` already exists and you try to merge in a config with `foo/my-rule-2` that is a different object?

> In other words, I don't think this proposal currently offers a solution to the issues discussed in eslint/eslint#3458 where two shareable configs depend on different versions of a plugin.

You are correct, this design doesn't address this problem because I don't think it's important to solve. This is an extreme edge case that I think we've spent way too much time and energy trying to solve. With this design, configs can export a namespace for any rules they depend on through `ruledefs` such that `eslint-config-airbnb` could have a `ruledefs` with a `airbnbreact` namespace (or `airbnb::react` namespace, it's up to them). Shareable configs could either export two different namespaces, for example, both `airbnb::react` and `react`, both pointing to the same object containing rules.

Reply:
> That's a good point. Would it make sense to throw an error if `foo/my-rule-2` already exists and you try to merge in a config with `foo/my-rule-2` that is a different object?

Unfortunately this solution wouldn't be robust because it would be relying on the details of how a package manager flattens packages, which could unexpectedly change without action by the end user.

<details>
<summary>More explanation</summary>

Suppose `config1` depends on versions `^1.0.0` of `eslint-plugin-foo`, and `config2` depends on versions `~1.0.0` of `eslint-plugin-foo`. If the current version is `1.0.0` when the end user creates their config, and the end user uses a package manager that happens to flatten packages, then `eslint-plugin-foo@1.0.0` will get installed as a dependency of both, so the user's config will work. However, if `eslint-plugin-foo@1.1.0` is released later, it might end up that `eslint-plugin-foo@1.1.0` would get installed for `config1` and `eslint-plugin-foo@1.0.0` would get installed for `config2`. As a result, the end user's config would unexpectedly break later on without any action taken by the end user. Tree flattening in package managers is an implementation detail, so I don't think it's a good idea to have a solution that relies on two packages being resolved to the same object.

</details>

---

> You are correct, this design doesn't address this problem because I don't think it's important to solve. This is an extreme edge case that I think we've spent way too much time and energy trying to solve.

Broadly speaking, I think it's necessary to consider edge cases like this when designing a system of interaction between third parties. Given how many users we have, even rare edge cases will occur frequently in absolute terms.

The reason this problem keeps coming up is that we're asking third parties who don't have any information about each other to introduce things into a global namespace without colliding with each other. This is impossible in general.

To make the problem tractable, it's necessary to impose constraints on names that third parties can use, either by:

* Giving the end user final say over what names can be introduced (like our current `peerDependencies` solution, which allows the end user to install and choose the version of their plugins)
* Automatically assigning new names when third parties introduce a name conflict (like the solution in https://github.com/eslint/rfcs/pull/5, which assigns new names using namespaces)
* Letting config authors manually assign new names when two of their dependencies conflict with each other (like "Solution 1" from https://github.com/eslint/rfcs/pull/9#discussion_r251960102)

It's important to me that we have a workable way to address the problem rather than just hoping it doesn't occur in practice. I would rather have a robust system that's slightly more difficult to use than a simple system that unexpectedly breaks 2% of the time.

Reply:
Just to be clear: I'm not saying I didn't consider this edge case, I'm saying I considered it and decided not to attempt to solve it directly. I don't believe solving this should be a requirement for this design, either.

As I've mentioned elsewhere, version conflicts between dependencies is really up to the package manager to solve. npm itself has complex logic to figure out which versions of which packages should be installed and in what directories to install them. If npm decides it cannot properly install two packages with different versions, then there's nothing ESLint can do to solve that.

There are also non-tech solutions to this problem. For example, if two configs conflict, the end user could manually copy what they wanted out of one shareable config into the place they want it. They could also manually install the plugin version that they want and stop using the shareable config altogether.

And just to set context, we have never seen this problem in the wild, we are only hypothesizing that this could happen (I believe I was the first to hypothesize this on https://github.com/eslint/eslint/issues/3458). I agree that doesn't mean the chance of it happening is 0%, but I do think the chance is sufficiently close to 0% that it's not something that should hold up this RFC.